File: JavaScript solutions/49. Group Anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  let ans = [];
  let p = [];

  if (strs.length === 0) {
    return ans;
  }

  for (let i = 0; i < strs.length; i++) {
    let s = strs[i].split("").sort().join("");
    p.push([s, i]);
  }

  p.sort((a, b) => a[0].localeCompare(b[0]));

  let res = [];
  let empty = [];

  res.push(strs[p[0][1]]);
  for (let i = 1; i < p.length; i++) {
    if (p[i][0] !== p[i - 1][0]) {
      ans.push([...res]);
      res = empty.slice();
      res.push(strs[p[i][1]]);
    } else {
      res.push(strs[p[i][1]]);
    }
  }
  ans.push([...res]);

  return ans;
};

File: C++ solutions/49. Group Anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>>ans;
        vector<pair<string,int>>p;
        if(strs.size()==0) return ans;

        for(int i=0;i<strs.size();i++){
            string s=strs[i];
            sort(s.begin(),s.end());
            p.push_back(make_pair(s, i));
        }
        sort(p.begin(),p.end()); 
        vector<string>res,empty;

        res.push_back(strs[p[0].second]);
        for(int i=1;i<p.size();i++){
            if(p[i].first!=p[i-1].first){
                ans.push_back(res);
                res=empty;
                res.push_back(strs[p[i].second]);
            }
            else res.push_back(strs[p[i].second]);
        }
        ans.push_back(res);
        return ans;
    }
};